Make RVD better

rynowak opened the pull request:
This change makes RouteValueDictionary into something incredible. Like a butterfly emerging from a chrysalis.

@Eilon are you terrified by this?

rynowak commented:
Sampling data from HelloWorldMvcCore (Techempower plaintext):
**Before**
![image](https://cloud.githubusercontent.com/assets/1430011/14623743/bc5e6432-0589-11e6-8920-72f791951d0c.png)


**After**
![image](https://cloud.githubusercontent.com/assets/1430011/14623755/cea4c35c-0589-11e6-8818-94705127cc13.png)


This change significantly speeds up routing by using a list instead of a dictionary. Operations hashing a string (ordinal ignore case) are taking too much time, and we usually store 3-4 values.

javiercn commented:
Looks good to me, Is it this [Design]?

rynowak commented:
> Is it this [Design]?

Yeah, I suppose it is. I'll need to add tests for the state transitions as well as the methods which we previously got for free from dictionary.

javiercn commented:
Design looks good to me. To summarize the changes you basically have a RVD that starts with an empty storage and upgrades to a list storage when a value is inserted, isn't it? 

Also, the improvement will be 4.36-0.99 (set_item) 3.37?

rynowak commented:
> Design looks good to me. To summarize the changes you basically have a RVD that starts with an empty storage and upgrades to a list storage when a value is inserted, isn't it?

Yep. Next phase will do even more crazy stuff

> Also, the improvement will be 4.36-0.99 (set_item) 3.37?

It's not quite that straightforward... **everything** in RVD should be faster now.  If you look at the time proportionally it's spending 5.45% in RVD methods before and 1.42% after..

Note that the time measurements are **not** directly comparable, you have to look at proportions of the total time spent in MVC+Routing+App which was 75,610ms before and 70,256ms after.

**Before**
![image](https://cloud.githubusercontent.com/assets/1430011/14644698/cc6f5498-0607-11e6-808c-2cc8283d2b03.png)


**After**
![image](https://cloud.githubusercontent.com/assets/1430011/14644711/da6c706c-0607-11e6-8763-43f11820ad4c.png)

dougbu commented on test/Microsoft.AspNetCore.Mvc.Routing.Abstractions.Tests/RouteValueDictionaryTests.cs:
```diff
@@ -156,21 +156,6 @@ public void CreateFromObject_CopiesPropertiesFromRegularType_WithIndexerProperty
         }
 
-        [Fact]
-        public void CreateFromObject_MixedCaseThrows()
```
We added this test because it could thoroughly confuse users if the code silently chose one value over the other. Or perhaps I did cause a customer issue...

Either way, we shouldn't lose this behaviour.

rynowak replied:
Yeah I agree, I think RVD will end up doing it's own cache of PropertyHelper objects. We don't want to lose this error message.

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -35,62 +38,51 @@ public RouteValueDictionary()
-                else
+
+                PropertyStorage propertyStorage;
+                if ((propertyStorage = dictionary._storage as PropertyStorage) != null)
```
Strange to use `else if` in the outer level and not here.

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -35,62 +38,51 @@ public RouteValueDictionary()
+            else
+            {
+                _storage = EmptyStorage.Instance;
                 return;
```
Most of the `return` statements in this constructor are just noise and make things longer. Either use early exits everywhere (removing the `else if` bits) or remove excess `return`s.

rynowak replied:
Changing this code to use more `if` and early exits.

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -137,17 +132,33 @@ public ICollection<string> Keys
-                return InnerDictionary.Keys;
+                Upgrade();
+
+                var list = ((ListStorage)_storage);
```
What are the outer parens for?

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -185,22 +214,36 @@ public void Add(string key, object value)
         {
-            EnsureWritable();
-            InnerDictionary.Clear();
+            Upgrade();
```
Should exit early if storage is currently `EmptyStorage`.

dougbu replied:
Applies to all delete-ish methods below e.g. `Remove()` should return `false` in `EmptyStorage` case.

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -185,22 +214,36 @@ public void Add(string key, object value)
         {
-            EnsureWritable();
-            return ((ICollection<KeyValuePair<string, object>>)InnerDictionary).Contains(item);
+            Upgrade();
```
Should exit early if storage is currently `EmptyStorage`. Particularly strange to upgrade for a read operation.

rynowak replied:
There's a few operations on dictionary/collection that aren't commonly used. Typically anything that's explicitly implemented isn't really **ever** going to be called. I'm opting for simplicity in those cases and trying to avoid writing 7-8 tests for every method.

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -293,114 +344,269 @@ public bool TryGetValue(string key, out object value)
+
+            object IEnumerator.Current => Current;
+
+            public void Dispose()
```
What is this method for?

rynowak replied:
MSDN will [explain it better than I can](https://msdn.microsoft.com/en-us/library/system.idisposable.dispose(v=vs.110).aspx)

dougbu replied:
Funny. What I was missing was that `IEnumerator<KeyValuePair<string, object>>` implements `IDisposable`.

dougbu replied:
:see_no_evil:

dougbu commented:
:watch:

Eilon commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -137,17 +132,33 @@ public ICollection<string> Keys
             {
-                EnsureWritable();
-                return InnerDictionary.Keys;
+                Upgrade();
```
Why always upgrade? Why not have the Storage do this?

Eilon replied:
Plus same comment in other similar scenarios. (e.g. generic Keys property below)

---
In reply to: [60490919](https://github.com/aspnet/Routing/pull/309#discussion_r60490919) [](ancestors = 60490919)

rynowak replied:
There's little payoff to trying to produce optimized implementations of everything on `IDictionary<,>` and `ICollection<>`. All of these extra implementations require their own tests.

The eventual goal of all of this is that an app using a 'normal' MVC action selection scheme or attribute routing doesn't need to `Upgrade` on any of the main code paths.

Eilon commented:
Overall looks good to me. Just some Upgrade() questions.

javiercn commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -35,62 +38,51 @@ public RouteValueDictionary()
         {
-            var otherDictionary = values as RouteValueDictionary;
-            if (otherDictionary != null)
+            RouteValueDictionary dictionary;
```
Do we intentionally want to allow null to be passed to this class? The code implies that we do, but i'm wondering if its intentional.

rynowak replied:
yes.

javiercn commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -35,62 +38,51 @@ public RouteValueDictionary()
+                _storage = EmptyStorage.Instance;
+                return;
+            }
+            else if ((keyValueEnumerable = values as IEnumerable<KeyValuePair<string, object>>) != null)
```
if you have to do return; at the end of each block I think it's clearer if you just write the code as a sequence of ifs, rather than doing if else all around, i think its clearer that way.

rynowak replied:
Why are any of you commenting on this in a [Design] PR?

javiercn replied:
Fair enough :smile:

javiercn commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -293,114 +344,269 @@ public bool TryGetValue(string key, out object value)
-        public struct Enumerator : IEnumerator<KeyValuePair<string, object>>
-        {
-            private readonly RouteValueDictionary _dictionary;
+            public override int Count => _inner.Count;
```
5 pranav points

rynowak commented:
Updated. 

I've added tests, and addressed most of the feedback. I'm in the middle of doing some more thorough testing to see what implementations need to be optimized. I've done a pass and made many of the operations that call upgrade check the count first and early exit where possible.

rynowak commented:
Removing the design tag on this. I verified with some profiling that we're not really ever upgrading needlessly. All of our upgrades happen because we're writing to the dictionary.

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -269,8 +314,30 @@ IEnumerator IEnumerable.GetEnumerator()
+                return false;
+            }
+
+            if (_storage.Count == 0)
```
Like the new `Count` checks. But I suspect you'll find the value hasn't changed since [line 317](https://github.com/aspnet/Routing/pull/309/files#diff-8a515f05675a880115920385ab225fcbR317) :trollface:

rynowak replied:
:+1:

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -293,115 +376,306 @@ public bool TryGetValue(string key, out object value)
-            public bool MoveNext()
+        internal class PropertyStorage : Storage
+        {
+            private static readonly PropertyCache _propertyCache = new PropertyCache();
```
Ick, another `static` cache on top of those in `PropertyHelper`. Any way we can avoid that? E.g. could expose a `virtual ValidateHelpers(PropertyHelper[] helpers)` method that `PropertyHelper.GetProperties(...)` calls before updating its cache.

rynowak replied:
Here's my proposal - let's add another overload/parameter to all of these https://github.com/aspnet/Common/blob/dev/src/Microsoft.Extensions.PropertyHelper.Sources/PropertyHelper.cs#L166 that accepts a `Func<PropertyInfo, PropertyHelper>`. 

This feature already exists in `PropertyHelper` - it's just not exposed as public.

rynowak replied:
Actually this won't work - we can't let anyone write to our cache because they aren't following the rules that we are.

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -293,115 +376,306 @@ public bool TryGetValue(string key, out object value)
+
+            private static void ValidatePropertyNames(Type type, PropertyHelper[] properties)
+            {
+                var names = new Dictionary<string, PropertyHelper>(StringComparer.OrdinalIgnoreCase);
```
Minor but this could use the `Comparer` property couldn't it?

rynowak replied:
No, not from a static method in a nested class. :see_no_evil:

dougbu commented on test/Microsoft.AspNetCore.Mvc.Routing.Abstractions.Tests/RouteValueDictionaryTests.cs:
```diff
@@ -4,44 +4,148 @@
 using System.Collections.Generic;
-using Microsoft.AspNetCore.Testing;
 using Xunit;
+using Microsoft.AspNetCore.Testing;
```
Was ordered correctly before.

dougbu commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -293,115 +376,306 @@ public bool TryGetValue(string key, out object value)
+        {
+            public static readonly EmptyStorage Instance = new EmptyStorage();
+
+            public override int Count => 0;
```
Suggest making the parameterless constructor `private` to keep future-Ryans honest.

rynowak replied:
sure

dougbu commented on test/Microsoft.AspNetCore.Mvc.Routing.Abstractions.Tests/RouteValueDictionaryTests.cs:
```diff
@@ -153,75 +269,1121 @@ public void CreateFromObject_CopiesPropertiesFromRegularType_WithIndexerProperty
-                dictionary.Add("Hi", "There");
             });
+
+            // Ignoring case to make sure we're not testing reflection's ordering.
```
:clap:

dougbu commented on test/Microsoft.AspNetCore.Mvc.Routing.Abstractions.Tests/RouteValueDictionaryTests.cs:
```diff
@@ -153,75 +269,1121 @@ public void CreateFromObject_CopiesPropertiesFromRegularType_WithIndexerProperty
+            var dict = new RouteValueDictionary(new List<KeyValuePair<string, object>>()
+            {
+                new KeyValuePair<string, object>("key", "value"),
+            });
```
Just
``` c#
new RouteValueDictionary
{
    { "key", "value" },
};
```

rynowak replied:
:+1:

dougbu commented:
:watch: hoping https://github.com/aspnet/Routing/pull/309#discussion-diff-60759496 can be improved.

rynowak commented:
updated

Eilon commented on src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs:
```diff
@@ -293,115 +368,310 @@ public bool TryGetValue(string key, out object value)
         }
 
-        private void EnsureWritable()
+        // Storage and it's subclasses are internal for testing.
```
Your write about this comment :trollface:

Eilon commented:
:shipit: from me.

dougbu commented:
:shipit:

rynowak commented:
btw this change takes MVC plaintext from 205k -> 220k /cc
